Add support for REDIS_URL environment variables.

robinvdvleuten opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20850 
| License       | MIT
| Doc PR        | -

In most PAAS solutions - Heroku for example - the Redis connection string is exposed as an environment variable like `REDIS_URL`. This PR adds support for those by allowing the following config;

```yml
framework:
    cache:
        default_redis_provider: "%env(REDIS_URL)%"
```

In the referenced issue there was some discussion about maybe a new config options like `default_redis_url`, but this makes it hard to check in the extension for the case that an user configured a custom _default_redis_provider_ and also has configured the default url.

robinvdvleuten committed:
Add support for REDIS_URL environment variables.

stof commented:
Can you add a test covering this ?

robinvdvleuten commented:
@stof sure! Just a sec then

nicolas-grekas commented:
That won't work: injecting a string (the REDIS_URL) in RedisAdapter is not allowed.
What you need to do is add a new config option, let's say default_redis_url, and use that to wire the REDIS_URL as a DSN in the same way that I suggested in https://github.com/symfony/symfony/issues/20850#issuecomment-266058449 (but programmatically in the DI extension.)

nicolas-grekas commented:
I should review twice before posting :) So this forces any `%env(...)%` to be understood as a DSN. OK then :)
Does this really work? I mean, env vars are usually seen as unique placeholders fro mDI ext., not a plain `%env()%` thing.
Adding a test case will show for sure.

robinvdvleuten commented:
@nicolas-grekas what I am trying to understand is why the service definition you gave in the ticket;

```yml
cache.default_redis_provider:
    class: Redis
    public: false
    factory: ["Symfony\\Component\\Cache\\Adapter\\RedisAdapter", "createConnection"]
    arguments:
      - "%env(REDIS_URL)%"
```

is different from this in PHP;

```php
$definition = new Definition(\Redis::class);
$definition->setPublic(false);
$definition->setFactory(array(RedisAdapter::class, 'createConnection'));
$definition->setArguments(array('%env(REDIS_URL)%'));
$container->setDefinition($name, $definition);
```

This should do the same in the compilation process right?

robinvdvleuten commented:
@nicolas-grekas ha you typed a response faster than I did. This approach worked in my application so I created this PR to get your opinion before putting any effort in the test cases. The `default_redis_provider` isn't testing at all btw.

nicolas-grekas commented:
When I apply your patch, I get the same:
```
[Symfony\Component\DependencyInjection\Exception\EnvParameterException]             
Incompatible use of dynamic environment variables "REDIS_URL" found in parameters.  
```

The correct patch might be more like:
```diff
--- a/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php
+++ b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php
@@ -106,7 +106,9 @@ class CachePoolPass implements CompilerPassInterface
      */
     public static function getServiceProvider(ContainerBuilder $container, $name)
     {
-        if (0 === strpos($name, 'redis://')) {
+        $container->resolveEnvPlaceholders($name, null, $usedEnvs);
+
+        if (0 === strpos($name, 'redis://') || $usedEnvs) {
             $dsn = $name;
 
             if (!$container->hasDefinition($name = md5($dsn))) {
```

Btw, I'd be OK on my side to have this as bugfix on 3.2.

robinvdvleuten committed:
Correctly resolve any possible environment variables.

robinvdvleuten committed:
Added test cases for the default redis provider.

robinvdvleuten commented:
@nicolas-grekas you are right, I was looking at the wrong thing. I've added the resolve call and updated the testcases to test both the default string and the env var.

robinvdvleuten committed:
Applied CS patch.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/yml/cache_env_var.yml:
```diff
@@ -0,0 +1,6 @@
+
+framework:
+    cache:
+        default_redis_provider: "%env(REDIS_URL)%"
```
missing EOL

robinvdvleuten replied:
Added the EOL

robinvdvleuten committed:
Added EOL.

nicolas-grekas commented:
:+1:

fabpot commented:
Thank you @robinvdvleuten.

nicolas-grekas commented:
This is the fastest ever submitted+reviewed+fixed+merged+released PR :)

dunglas commented:
However, it seems that it breaks tests: https://travis-ci.org/symfony/symfony/jobs/183649383

dunglas commented:
Sorry, this isn't related: symfony/symfony#20906

chalasr commented:
I do think it's related since the test only fails in its `redis` version with a custom config, it succeeds otherwise
